solr/core/src/java/org/apache/solr/search/join/XCJFQuery.java

```diff
@@ -0,0 +1,380 @@
+
+    TermsEnum termsEnum;
+    BytesRefBuilder bytes;
+    PostingsEnum postingsEnum;
+
+    FixedBitSet bitSet;
+
+    public TermsJoinKeyCollector(FieldType fieldType, Terms terms, SolrIndexSearcher searcher) throws IOException {
+      this.fieldType = fieldType;
+      this.searcher = searcher;
+
+      termsEnum = terms.iterator();
+      bytes = new BytesRefBuilder();
+
+      bitSet = new FixedBitSet(searcher.maxDoc());
+    }
+
+    @Override
+    public void collect(Object value) throws IOException {
+      fieldType.readableToIndexed((String) value, bytes);
+      if (termsEnum.seekExact(bytes.get())) {
+        postingsEnum = termsEnum.postings(postingsEnum, PostingsEnum.NONE);
+        bitSet.or(postingsEnum);
+      }
+    }
+
+    @Override
+    public DocSet getDocSet() throws IOException {
+      if (searcher.getIndexReader().hasDeletions()) {
+        bitSet.and(searcher.getLiveDocSet().getBits());
+      }
+      return new BitDocSet(bitSet);
+    }
+  }
+
+  private class PointJoinKeyCollector extends GraphPointsCollector implements JoinKeyCollector {
+
+    SolrIndexSearcher searcher;
+
+    public PointJoinKeyCollector(SolrIndexSearcher searcher) {
+      super(searcher.getSchema().getField(toField), null, null);
+      this.searcher = searcher;
+    }
+
+    @Override
+    public void collect(Object value) throws IOException {
+      if (value instanceof Long || value instanceof Integer) {
+        set.add(((Number) value).longValue());
+      } else {
+        throw new UnsupportedOperationException("Unsupported field type for XCJFQuery");
+      }
+    }
+
+    @Override
+    public DocSet getDocSet() throws IOException {
+      Query query = getResultQuery(searcher.getSchema().getField(toField), false);
+      if (query == null) {
+        return DocSet.EMPTY;
+      }
+      return DocSetUtil.createDocSet(searcher, query, null);
+    }
+  }
+
+  private class XCJFQueryWeight extends ConstantScoreWeight {
+
+    private SolrIndexSearcher searcher;
+    private ScoreMode scoreMode;
+    private Filter filter;
+
+    public XCJFQueryWeight(SolrIndexSearcher searcher, ScoreMode scoreMode, float score) {
+      super(XCJFQuery.this, score);
+      this.scoreMode = scoreMode;
+      this.searcher = searcher;
+    }
+
+    private String createHashRangeFq() {
+      if (routedByJoinKey) {
+        ClusterState clusterState = searcher.getCore().getCoreContainer().getZkController().getClusterState();
+        CloudDescriptor desc = searcher.getCore().getCoreDescriptor().getCloudDescriptor();
+        Slice slice = clusterState.getCollection(desc.getCollectionName()).getSlicesMap().get(desc.getShardId());
+        DocRouter.Range range = slice.getRange();
+
+        // In CompositeIdRouter, the routing prefix only affects the top 16 bits
+        int min = range.min & 0xffff0000;
+        int max = range.max | 0x0000ffff;
+
+        return String.format(Locale.ROOT, "{!hash_range f=%s l=%d u=%d}", fromField, min, max);
```

Comment:
How come you went with this hash_range filter on the remote handler instead of adding a _route_parameter on the query itself?

I'm new to Solr code so not sure if _route_ parameter is feasible here; or if you just had other purposes/constraints.

Reply:
@zkendall This repo is the old 8.x line of Solr, the latest work is in github.com/apache/solr.    Also, you may get more traction on your question on the solr dev mailing list...

Reply:
The route parameter would route the query to one specific shard.  In this scenario, we want the query to execute on every shard, however, only return the joinkeys from those shards within the given hash range.